src/gwt/acesupport/acemode/sweave_background_highlighter.js

```diff
@@ -136,29 +136,58 @@ define("mode/sweave_background_highlighter", function(require, exports, module)
          }
       };
 
+      this.$clearMarkers = function() {
+         for (var i = 0; i < this.$markers.length; i++)
+            this.$session.removeMarker(this.$markers[i]);
+         this.$markers = [];
+      };
+
       this.$syncMarkers = function(startRow, rowsChanged) {
-         var dontStopBeforeRow =
-               (typeof(rowsChanged) == 'undefined' ? this.$doc.getLength()
-                                                   : startRow + rowsChanged);
 
-         var endRow = this.$doc.getLength() - 1;
+         // Determine how many markers need to be updated. We
+         // attempt to be conservative and only 
+         var endRow;
+         if (rowsChanged == null)
+            endRow = this.$doc.getLength() - 1;
+         else
+            endRow = startRow + rowsChanged;
+
+         if (endRow >= this.$doc.getLength())
+            endRow = this.$doc.getLength() - 1;
+
+         // If the initial state is an 'end' state, we need to update
+         // all markers later on in the document, as we may have just
+         // closed a chunk.
+         var state = this.$rowState[startRow];
+         if (state === TYPE_END)
+            endRow = this.$doc.getLength() - 1;
+
          for (var row = startRow; row <= endRow; row++) {
-            var foreign = this.$rowState[row] != TYPE_TEXT;
-            if (!!foreign != !!this.$markers[row]) {
```

Comment:
I don't know what this line is doing (too clever for its own good?) and was encountering problems with identifying the 'start' of a chunk so I ended up re-writing this block of code.